alliance/share/man/man1/genlib.1
.\"   $Id: genlib.1,v 1.1 1999/05/31 17:30:13 alliance Exp $
.\" @(#)Labo.l 0.0 90/22/08 UPMC; Author: Frederic Petrot
.TH GENLIB 1 "October 1, 1997" "ASIM/LIP6" "ALLIANCE USER COMMANDS"
.SH NAME
genlib \- Procedural design language based upon C, along with its
compiler 
.so man1/alc_origin.1
.SH DESCRIPTION
\fBgenlib\fP is a set of \fIC\fP functions dedicated to procedural generation
purposes.
From a user point of view, \fBgenlib\fP is a circuit's description language
that allows standard \fIC\fP programming flow control, variable use, and 
specialized functions in order to handle vlsi objects.
.br
Based upon the \fIAlliance\fP \fBmbk\fP data structures, the \fBgenlib\fP
language gives the user the ability to describe both netlist and
layout views, thus allowing both standard cell and full custom approachs.
.br
.TP 20
Netlist capture :
\- it is a hierachical structural description of
a circuit in terms of connectors (I/Os), signals (nets), and instances.
.br
The function calls used to handle the netlist view are
.RS
	DEF_LOFIG(3)
.br
	SAVE_LOFIG(3)
.br
	LOINS(3)
.br
	LOCON(3)
.br
	LOSIG(3)
.br
	FLATTEN_LOFIG(3)
.br
Some facilities, in order to create vectors are also available :
.br
	BUS(3)
.br
	ELM(3)
.RE
.TP 20
Standard cell placement
The following functions allows to define a placement file for a standard cell
design.
This file can be used by the standard cell router \fBscr\fP(1).
.RS
	DEF_PHSC(3)
.br
	SAVE_PHSC(3)
.br
	SC_PLACE(3)
.br
	SC_RIGHT(3)
.br
	SC_TOP(3)
.br
	SC_LEFT(3)
.br
	SC_BOTTOM(3)
.br
.RE
.TP 20
Full custom symbolic layout
Those functions are dedicated to optimized full custom procedural layout.
In order to provide some process independance, \fIAlliance\fP uses a
symbolic layout approach (fixed grid without compaction.)
.br
The symbolic objects are segments (wires), vias (contacts), connectors (I/Os),
references and instances.
For more informations, see \fBphseg\fP(1), \fBphvia\fP(1), \fBphcon\fP(1),
\fBphref\fP(1), \fBphins\fP(1), and \fBalc\fP(1).
.br
	DEF_PHFIG(3)
.br
	SAVE_PHFIG(3)
.br
	DEF_AB(3)
.br
	DEF_PHINS(3)
.br
	PHCON(3)
.br
	COPY_UP_CON(3)
.br
	COPY_UP_CON_FACE(3)
.br
	COPY_UP_ALL_CON(3)
.br
	PHSEG(3)
.br
	COPY_UP_SEG(3)
.br
	THRU_H(3)
.br
	THRU_V(3)
.br
	THRU_CON_H(3)
.br
	THRU_CON_V(3)
.br
	WIRE1(3)
.br
	WIRE2(3)
.br
	WIRE3(3)
.br
	PHVIA(3)
.br
	PLACE(3)
.br
	PLACE_RIGHT(3)
.br
	PLACE_TOP(3)
.br
	PLACE_LEFT(3)
.br
	PLACE_BOTTOM(3)
.br
	PLACE_ON(3)
.br
	PHREF(3)
.br
	COPY_UP_REF(3)
.br
	COPY_UP_ALL_REF(3)
.br
	PLACE_VIA_REF(3)
.br
	PLACE_CON_REF(3)
.br
	PLACE_SEG_REF(3)
.br
	FLATTEN_PHFIG(3)
.br
	GET_REF_X(3)
.br
	GET_REF_Y(3)
.br
	GET_CON_X(3)
.br
	GET_CON_Y(3)
.br
	HEIGHT(3)
.br
	WIDTH(3)
.RE
.LP
In order to have information about each of these functions, use the online
documentation with \fBman\fP(1), as in ``man function-name''.
.br
.LP
It is strongly recommended to read some books
on C programming, in order to take full advantage of the C flow control
possibilities, as it may greatly reduce the size of a \fBgenlib\fP source code.
.SH ENVIRONMENT VARIABLES
.TP 20
the variables are :
.ta 25n
\fIname\fP	\fIdefault value\fP
.br
MBK_IN_LO(1)	al
.br
MBK_OUT_LO(1)	al
.br
MBK_IN_PH(1)	ap
.br
MBK_OUT_PH(1)	ap
.br
MBK_CATA_LIB(1)	.
.br
MBK_WORK_LIB(1)	.
.br
MBK_CATAL_NAME(1)	CATAL
.LP
see the corresponding manual pages for further informations.
.br
In order to compile and execute a genlib file, one has to call \fBgenlib\fP
with one argument, that is the \fBgenlib\fP source file.
The source file must have a .c extension, but the extension should not be
mentionned on the command line.
.PP
The names used in genlib, as arguments to genlib functions, should be
alphanumerical, including the underscore.
They also are not case sensitive, so \fBVDD\fP is equivalent to \fBvdd\fP.
Vectorized connectors or signal can be declareds using the \fB[n:m]\fP
construct.
.SH OPTIONS
Two options can alter the behavior of \fBgenlib\fP : 
.TP 20
-v
displays everything that is to be done during the compilation and
execution process: verbose mode.
.TP
-k
keeps the executable along with the compilation Makefile after completion.
.SH EXAMPLES
\fBgenlib\fP -v amd2901
.br
Compiles and runs a \fBgenlib\fP source file called amd2901.c
.SH SEE ALSO
.BR mbk (1),
.BR alc (1),
.BR scr (1),
.BR bgd (1),
.BR bsg (1),
.BR grog (1),
.BR rsa (1).

.SH DIAGNOSTICS
.RS
\fBCompilation aborted\fP
.RE
This indicates either a \fIC\fP syntaxic error, most likely, or a
misinstallation of the \fBmbk\fP package of the \fIAlliance\fP CAD system,
if the error is \fBcannot find file genlib.h\fB.
.RS
\fBLink aborted\fP
.RE
This indicates a
misinstallation of the \fBmbk\fP package of the \fIAlliance\fP CAD system,
see your system administrator for details.
.PP
Many errors may occur while executing the source file, so refer to the proper
genlib function manual for more.

.so man1/alc_bug_report.1



alliance/CHANGES
# Alliance VLSI CAD System
#
# Home page          : http://www-asim.lip6.fr/alliance/
# E-mail support     : mailto:alliance-support@asim.lip6.fr
# ftp sites          : ftp://ftp.lip6.fr/lip6/softs/alliance/
#                   or ftp://ftp-asim.lip6.fr/pub/alliance/
#
# $Id: CHANGES,v 1.3 2000/01/19 18:35:30 czo Exp $

--------------------------------------------------------------------------------

ALLIANCE release 4.0 (15/10/99)

1/ New Standard Cells Library : sxlib
   sxlib allows multi-layer over cell routing
   and all transitions are now on rising edge (man sxlib)

2/ Improved Tools :
   - asimut is now a temporal logic simulator. 
     You *must* modify your old pattern files to allow delays (man asimut)
   - syf can now handle vbe statements in .fsm
   - dreal : New viewing options ( like -install for private colormap)
   - graal : New viewing options ( like -install for private colormap)
             Input scale changeable, Hierarchy browser, real view (.cif/.gds)
   
3/ New Tools :
   algue : ALliance Graphic User Environement
   ali   : gives information on Alliance Environnement
   b2f   : FSM Abstraction from .vbe
   k2f   : Translate kiss2 format <-> FSM format
   vasy  : Translate .vhd format (Synopsys) <-> .vbe format 
   xsch  : Netlist viewer
   xfsm  : FSM viewer
   xpat  : patterns (with delay) viewer


--------------------------------------------------------------------------------

ALLIANCE release 3.5 (19/06/98)

1/ Lots of improvments and bug fixes

2/ AP : Now supports ten layers of metal

3/ AL : New support for RC net in losig. Parsers al, spice, vti modified.
   Modified Data Structures :
    - lotrs : added field BULK TRNAME
    - locon : added field PNODE
    - losig : added field PRCN, CAPA deleted
    
   Modified functions (the CAPA field has moved) :
        addlosig, addlotrs, dellosig

4/ RCN : New library for RC support : Resistances on losig and capcitances
   beetwen losig.
   
--------------------------------------------------------------------------------

ALLIANCE release 3.2c (20/03/98)

1/ Ap format :

   Alliance ap format now supports 
   - half grid spacing
   - connectors somewhere else than abutment box.

2/ VERILOG driver :

   Support added for Alliance-Cadence Toolbox

3/ MBK_SCALE_X now defaults to 100

--------------------------------------------------------------------------------

ALLIANCE release 3.2b (15/12/97)

1/ Easy install

   A "configure" script is now available to configure
   Alliance on any UNIX system

2/ New driver :

   Support to VERILOG netlist as been added.
   Only driver exits. This means you can save your 
   netlists to VERILOG format 'vlg'

3/ New names :

  - Logic has been splited in 3 parts,
    . bop   : boolean optimizer (logic -o) 
    . scmap : Std cell mapping (logic -s) 
    . c4map : CCCC mapping (logic -c) 

  - Desb is replaced by yagle

  - Alligator is replaced by fpmap (X4000)

  - Netoptim is replaced glop

4/ Cells libraries

   The tree of directories containing the cells 
   libraries has been simplified

--------------------------------------------------------------------------------

ALLIANCE release 3.2 (17/05/97)


1/ GRAPHICAL PATTERN VIEWER

   In order to see the patterns resulting from a simulation,
   the XPAT tools has been developped.

2/ GRAPHICAL FSM VIEWER

   In order to see the state's graph of an FSM the XFSM tools has 
   been developped.

3/ RECTANGLE LAYOUT VIEWER

   Now, DREAL is also a real layout editor.

--------------------------------------------------------------------------------

ALLIANCE release 3.0 (17/05/94)

1/ FPGA SYNTHESIS

   A logic synthesis tools that maps on FPGA is now available.
   It works for X3000 devices

2/ FLOOR-PLAN ROUTING

   Binaries of the CHEOPS router from BULL are available for sparc.
   
3/ TIMING ANALYSIS

   The static timing analisys tools TAS is finally available.
   It can be targetted to several processes though the use of a technological
   file suffixed `elp'.

4/ RECTANGLE LAYOUT VIEWER

   In order to see the layout resulting from a symbolic to real translation,
   the DREAL tools has been developped.
   
--------------------------------------------------------------------------------
ALLIANCE release 2.0 (14/02/94) versus ALLIANCE 1.2

1/ SYMBOLIC LAYOUT EDITOR

  The symbolic layout editor ALC has been replaced by GRAAL.
  GRAAL provides the same functionnalities than ALC, but is much
  more reliable. GRAAL support both CMOS and GaAs symbolic layout.

> man graal

2/ DESIGN RULE CHECKER

  The Design Rule Checker VERSATIL has been replaced by DRUC.
  DRUC provides the same functionnalities than VERSATIL.
  A hierarchical version will be distributed in the next ALLIANCE release.

> man druc

3/ LOGIC SYNTHESIS

  The logic synthesis tool LOGIC has been strongly improved.
  The new tool NETOPTIM is a gate-level net-list optimiser
  that allows to minimize delays in a synthesized gate net-list.
  The Finite-state-machine synthesizer SYF allows to describe and
  synthesize high complexity FSM (more than 100 states)
  It is possible to describe hierachical FSM using stack  (subroutines).

> man logic
> man syf
> man netoptim

4/ DATA-PATH COMPILER

  FPGEN is a data-path compiler using a dedicated macro-cells library.
  DPR is the place and route tool that creates optimized data-path blocks
  from the gate net-list generated by FPGEN.

> man fpgen
> man dpr

5/ PARAMETERIZED MACRO-CELLS

  Six parametrized generators are part of this release:

> man rsa                # fast adder generator
> man bsg                # barrel shifter generator
> man amg                # multiplier generator
> man rfg                # register file generator
> man grog               # high speed ROM generator
> man rage               # static RAM generator

6/ PROCEDURAL LAYOUT DEBUGGER

  The graphic debugger GENVIEW allows to debug custom blocks
  described with the procedural language GENLIB.

alliance/src/genlib/src/genlib.sh
#!/bin/sh
#
# $Id: genlib.sh,v 1.7 2002/09/30 16:20:23 czo Exp $
#                                                                        
# /------------------------------------------------------------------\
# |                                                                  |
# |        A l l i a n c e   C A D   S y s t e m                     |
# |                      D p G e n                                   |
# |                                                                  |
# |  Author    :                      Jean-Paul CHAPUT               |
# |  E-mail    :         alliance-users@asim.lip6.fr               |
# | ================================================================ |
# |  sh script :         "./dpgen/dpgen.sh"                          |
# | **************************************************************** |
# |  U p d a t e s                                                   |
# |                                                                  |
# \------------------------------------------------------------------/
#


# /------------------------------------------------------------------\
# |                                                                  |
# |                      Functions Definitions                       |
# |                                                                  |
# \------------------------------------------------------------------/


# --------------------------------------------------------------------
# Function  :  `print_usage()'.

 print_usage()
 {
   echo ""
   echo "  usage: genlib [-cdlknv] [--keep-exec] [--no-exec] [--no-rm-core] \\"
   echo "               [--keep-log] [--verbose] \\"
   echo "               <program> [-e [args]]"
   echo ""
   echo "  Options:"
   echo "    [-c|--no-rm-core]    : Don't remove core file."
   echo "    [-l|--keep-log]      : Keep log file (.grr) after execution."
   echo "    [-k|--keep-exec]     : Keep binary after execution."
   echo "    [-m|--keep-makefile] : Keep makefile after execution."
   echo "    [-n|--no-exec]       : Do not execute the binary."
   echo "    [-v|--verbose]       : Issue more informations."
   echo "    [-e [args]]          : [args] are passed to the binary."
   echo ""
 }


# --------------------------------------------------------------------
# Function  :  `vecho()'.

 vecho()
 { 
   ECHOFLAGS="" 
   if [ $1 -gt $VERBOSE ]; then return 0; fi
   if [ "$2" = "-n" ]; then
     ECHOFLAGS="-n"; shift;
   fi
   echo $ECHOFLAGS "$2" 2>&1 | tee -a $CSOURCE.grr
 }


# /------------------------------------------------------------------\
# |                                                                  |
# |                  Main Part of the Shell Script                   |
# |                                                                  |
# \------------------------------------------------------------------/

 if [ -z "$ALLIANCE_TOP" ]; then
   alcbanner "GenLib" "4.0" "Procedural Generation Language" "1991"
 else
   ${ALLIANCE_TOP}/bin/alcbanner "GenLib" "4.0" "Procedural Generation Language" "1991"
 fi



# --------------------------------------------------------------------
# Process the command line.

# If no arguments, print usage.
 if [ $# -eq 0 ]; then print_usage; exit 0; fi


  KEEP_BIN="n"
  KEEP_LOG="n"
 KEEP_CORE="n"
 KEEP_MAKEFILE="n"
   VERBOSE=0
   EXECUTE="y"
 ARGUMENTS=""
   PROGRAM="not set"
     DEVEL="n"


# Parse the arguments.
 while [ $# -gt 0 ]; do

   case $1 in

     --no-rm-core) KEEP_CORE="y";;
     --keep-makefile)  KEEP_MAKEFILE="y";;
     --keep-exec)  KEEP_BIN="y";;
     --keep-log)   KEEP_LOG="y";;
     --no-exec)    EXECUTE="n";;
     --verbose)    VERBOSE=1;;
     --devel)      DEVEL="y";;

     -*) NB=2; CH=`echo $1 | cut -c$NB`
	 while [ "$CH" != "" ]; do

	   case $CH in

             m) KEEP_MAKEFILE="y";;
             c) KEEP_CORE="y";;
             k) KEEP_BIN="y";;
             l) KEEP_LOG="y";;
             n) EXECUTE="n";;
             v) VERBOSE=1;;
             d) DEVEL="y";;
             e) echo "genlib:warning: misplaced -e option.";
                print_usage; exit 1;;
             *) echo "genlib:warning: Invalid option \`$CH' (ignored)."
                print_usage;;

	   esac

           NB=`expr $NB + 1`
           CH=`echo $1 | cut -c$NB`

         done;;

     *)            PROGRAM=$1; shift; break;;

   esac

   shift

 done


# Parse optionals program arguments.
 if [ $# -gt 0 ]; then
   if [ "$1" != "-e" ]; then

     echo "genlib:warning: ignored \`$1' parameter after program name."
     print_usage

   else

     shift; ARGUMENTS="$*"

   fi
 fi


# Check arguments coherency.
 if [ "$PROGRAM" = "not set" ]; then

   echo "genlib:error: Missing <program> argument."
   print_usage; exit 1;

 fi

 CSOURCE=`basename $PROGRAM .c`


# Removing previously generated files.
 rm -f $CSOURCE.grr $CSOURCE.o > /dev/null 2>&1


# --------------------------------------------------------------------
# Print the current MBK environment.

 if [ ! -z "$ALLIANCE_TOP" ]; then ALLIANCE_TOP=$ALLIANCE_TOP; fi

 vecho 1    "  o  Current Alliance environment:"
 vecho 1    "     - ALLIANCE_TOP    : ${ALLIANCE_TOP-not set}"
 vecho 1    "  o  Current MBK environment:"
 vecho 1    "     - MBK_IN_LO       : ${MBK_IN_LO-not set}"
 vecho 1    "     - MBK_OUT_LO      : ${MBK_OUT_LO-not set}"
 vecho 1    "     - MBK_IN_PH       : ${MBK_IN_PH-not set}"
 vecho 1    "     - MBK_OUT_PH      : ${MBK_OUT_PH-not set}"
 vecho 1    "     - MBK_WORK_LIB    : ${MBK_WORK_LIB-not set}"
 vecho 1 -n "     - MBK_CATA_LIB    : "

 if [ -z "$MBK_CATA_LIB" ]; then

   vecho 1 "not set"

 else

   STRING=`echo ${MBK_CATA_LIB} |                            \
             awk 'BEGIN { FS=":"; }                          \
                  { for( i=1; i<=NF; i++) {                  \
                      printf("%s\n", $i);                    \
                      if (i < NF)                            \
                        printf("                         "); \
                    }                                        \
                  }' -`

   vecho 1 "$STRING"

 fi

 vecho 1    "     - MBK_CATAL_NAME  : ${MBK_CATAL_NAME-not set}"
 vecho 1    "     - MBK_VDD         : ${MBK_VDD-not set}"
 vecho 1    "     - MBK_VSS         : ${MBK_VSS-not set}"
 vecho 1    "     - RDS_TECHNO_NAME : ${RDS_TECHNO_NAME-not set}"
 vecho 1    ""


# --------------------------------------------------------------------
# Development hacks.

 if [ "$DEVEL" = "y" ]; then
   ALLIANCE_INSTALL_DIR="__ALLIANCE_INSTALL_DIR__"
   ALLIANCE_DEVEL_INCLUDE="-I${ALLIANCE_INSTALL_DIR}/include"
   ALLIANCE_DEVEL_LIB="-L${ALLIANCE_INSTALL_DIR}/lib"
   export ALLIANCE_INSTALL_DIR ALLIANCE_DEVEL_INCLUDE ALLIANCE_DEVEL_LIB

   vecho 0    "  o  Switching to development environment :"
   vecho 0    "     - ALLIANCE_INSTALL_DIR : ${ALLIANCE_INSTALL_DIR-not set}"
 fi


# --------------------------------------------------------------------
# Build and execute the Makefile.

 vecho 1 "  o  Program generation."
 vecho 1 "     - Creating Makefile."

 MAKEFILE="Makefile-$$"


# --------------------------------------------------------------------
# Start of the Makefile text.

cat << EOF > $MAKEFILE
  ALLIANCE_INCLUDE = $ALLIANCE_TOP/include
  ALLIANCE_LIB     = $ALLIANCE_TOP/lib

  LOCAL_ALC_IDIR = $ALLIANCE_DEVEL_INCLUDE -I\$(ALLIANCE_INCLUDE)
  LOCAL_ALC_DEFS = 
  LOCAL_ALC_INCS = \$(LOCAL_ALC_IDIR) \$(LOCAL_ALC_DEFS)

  LOCAL_ALC_LDIR = $ALLIANCE_DEVEL_LIB -L\$(ALLIANCE_LIB)
  LOCAL_ALC_LVBE = -lBvl \\
                   -lBeh \\
                   -lBhl \\
                   -lLog
  LOCAL_ALC_LMGN = -lMgn
  LOCAL_ALC_LMUT = -lMut
  LOCAL_ALC_LMLO = -lMlu \\
                   -lMlo \\
                   -lMal \\
                   -lMcl \\
                   -lMgl \\
                   -lMel \\
                   -lMhl \\
                   -lMsl \\
                   -lMvg \\
                   -lRcn
  LOCAL_ALC_LMPH = -lMpu \\
                   -lMph \\
                   -lMap \\
                   -lMcp \\
                   -lMmg
  LOCAL_ALC_LIBS = \$(LOCAL_ALC_LDIR) \\
                   \$(LOCAL_ALC_LVBE) \\
                   \$(LOCAL_ALC_LMGN) \\
                   \$(LOCAL_ALC_LMLO) \\
                   \$(LOCAL_ALC_LMPH) \\
                   \$(LOCAL_ALC_LMUT)

  LOCAL_STD_LIBS = -lm

#  C Flags for normal compilation.
 LOCAL_CFLAGS = \$(CFLAGS) -c -g
 LOCAL_OFLAGS = \$(OFLAGS) -o


all: $CSOURCE

$CSOURCE: $CSOURCE.o
	\$(CC) \$(LOCAL_OFLAGS) \$@ \$? \$(LOCAL_ALC_LIBS) \$(LOCAL_STD_LIBS)
	rm -f \$?

$CSOURCE.o: $CSOURCE.c
	\$(CC) \$(LOCAL_CFLAGS) \$(LOCAL_ALC_INCS) \$?


clean:
	rm -f $CSOURCE $CSOURCE.o


EOF


# --------------------------------------------------------------------
# End of the Makefile text.


 vecho 1 "     - Running Makefile."

 make -f $MAKEFILE ALLIANCE_TOP=$ALLIANCE_TOP \
         CSOURCE=$CSOURCE clean all >> $CSOURCE.grr 2>&1

 exit_code=$?

 if [ $exit_code -ne 0 ]; then

   if [ $KEEP_MAKEFILE != "y" ]; then rm -f $MAKEFILE; fi

   echo "genlib:error: Compilation failed."
   echo "error messages:"
   cat $CSOURCE.grr
   exit $exit_code

 else

   if [ "$EXECUTE" = "y" ]; then

     vecho 1 "  o  Execute program..."
     ./$CSOURCE $ARGUMENTS; exit_code=$?

     vecho 1 "     - Exit code := $exit_code."
     if [ $exit_code -ne 0 ]; then
       if [ $KEEP_MAKEFILE != "y" ]; then rm -f $MAKEFILE; fi
       echo "genlib:error: Execution failed."
     fi

     if [ "$KEEP_CORE" != "y" ]; then rm -f core >> $CSOURCE.grr 2>&1; fi

   else
     vecho 1 "  o  Execution step skipped."
     exit_code=0
   fi

 fi


 vecho 1 ""
 vecho 1 "  o  Remove temporary file."

 rm -f $MAKEFILE

 if [ "$KEEP_BIN"  != "y" ]; then rm -f $CSOURCE >> $CSOURCE.grr 2>&1; fi

 if [ \( $exit_code -eq 0 \) -a \( "$KEEP_LOG" != "y" \) ]; then
   rm -f $CSOURCE.grr > /dev/null 2>&1
 fi

 echo ""


 exit $exit_code
